tee-worker/litentry/core/vc-task/receiver/src/lib.rs

```diff
@@ -154,30 +174,158 @@ where
 	N::MetadataType: NodeMetadataTrait,
 {
-	let start_time = Instant::now();
-	let enclave_shielding_key = context
-		.shielding_key
-		.retrieve_key()
-		.map_err(|e| format!("Failed to retrieve shielding key: {:?}", e))?;
-	let tcs = request
+	let request = &mut vc_req.request;
+	let sender = &vc_req.sender;
+	let enclave_shielding_key = match context.shielding_key.retrieve_key() {
+		Ok(value) => value,
+		Err(e) => {
+			send_vc_response(
+				sender,
+				context,
+				Err(format!("Failed to retrieve shielding key: {:?}", e)),
+				0u8,
+				1u8,
+			);
+			return
+		},
+	};
+	let tcs = match request
 		.decrypt(Box::new(enclave_shielding_key))
 		.ok()
 		.and_then(|v| TrustedOperation::<TrustedCallSigned, Getter>::decode(&mut v.as_slice()).ok())
 		.and_then(|top| top.to_call().cloned())
-		.ok_or_else(|| "Failed to decode payload".to_string())?;
-
+	{
+		Some(tcs) => tcs,
+		None => {
+			send_vc_response(
+				sender,
+				context,
+				Err("Failed to decode request payload".to_string()),
+				0u8,
+				1u8,
+			);
+			return
+		},
+	};
 	let mrenclave = match context.ocall_api.get_mrenclave_of_self() {
 		Ok(m) => m.m,
-		Err(_) => return Err("Failed to get mrenclave".to_string()),
+		Err(_) => {
+			send_vc_response(sender, context, Err("Failed to get mrenclave".to_string()), 0u8, 1u8);
+			return
+		},
 	};
+	if !tcs.verify_signature(&mrenclave, &request.shard) {
+		send_vc_response(sender, context, Err("Failed to verify sig".to_string()), 0u8, 1u8);
+		return
+	}
+	if let TrustedCall::request_vc(..) = tcs.call {
+		let response = process_single_request(
+			request.shard,
+			context.clone(),
+			extrinsic_factory,
+			node_metadata_repo,
+			tc_sender,
+			tcs.call.clone(),
+		);
+		send_vc_response(sender, context, response, 0u8, 1u8);
+	} else if let TrustedCall::request_batch_vc(signer, who, assertions, maybe_key, req_ext_hash) =
+		tcs.call
+	{
+		// Filter out duplicate assertions
+		let mut seen: HashSet<H256> = HashSet::new();
+		let mut unique_assertions = Vec::new();
+		for assertion in assertions.into_iter() {
+			let hash = H256::from(blake2_256(&assertion.encode()));
+			if seen.insert(hash) {
+				unique_assertions.push(Some(assertion));
+			} else {
+				unique_assertions.push(None);
+			}
+		}
+
+		let pool = ThreadPool::new(4);
+		let len = unique_assertions.len() as u8;
+		for (idx, assertion) in unique_assertions.iter().enumerate() {
+			let context_pool = context.clone();
+			let sender_clone = sender.clone();
+
+			if let Some(assertion) = assertion {
+				let new_call = TrustedCall::request_vc(
+					signer.clone(),
+					who.clone(),
+					assertion.clone(),
+					maybe_key,
+					req_ext_hash,
+				);
+				let extrinsic_factory_pool = extrinsic_factory.clone();
+				let node_metadata_repo_pool = node_metadata_repo.clone();
+				let tc_sender_pool = tc_sender.clone();
+				let shard = request.shard;
+
+				pool.execute(move || {
+					let response = process_single_request(
```

Comment:
There is an edge case here, That is if the `identity-graph` is empty, We will try to create `link-identity` trusted-call for all of these requests and submit, So then each response will have `mutated_id_graph`. 
CC: @Kailai-Wang

Reply:
Good catch.

I'd argue the impact is insignificant though - it can be mitigated on the client side @jonalvarezz . I don't know how the sdk currently deals with it, but in general you could igore the `pre_mutated_id_graph` if the local hash is equal to `pre_id_graph_hash`.

We could adjust the parachain side so that only the last response contain meaningful values, but again what if that vc request ends up being an error? The client would never get a response with these fields, which doesn't seem to be right either. So I'd say let's leave it as it is.

One improvement that I can see is we only need to submit `TrustedCall::maybe_create_id_graph` once per batch. In general, all batch level work should be extracted from `process_single_request`, so it's like:
```
- prepocessing
- n * vc request handling
- postprocessing
```

Please create an issue for that

Reply:
> I'd argue the impact is insignificant though - it can be mitigated on the client side @jonalvarezz .

Furthermore, `pre_mutated_id_graph` isn't used so far. IDH rely on the `get_id_graph_hash`.